src/category_theory/limits/shapes/zero.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.shapes.terminal
+
+universes v u
+
+open category_theory
+
+namespace category_theory.limits
+
+variables (C : Type u) [𝒞 : category.{v+1} C]
+include 𝒞
+
+/-- A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. -/
+class has_zero_morphisms :=
+[has_zero : Π X Y : C, has_zero (X ⟶ Y)]
+(comp_zero' : ∀ {X Y : C} (f : X ⟶ Y) (Z : C), f ≫ (0 : Y ⟶ Z) = (0 : X ⟶ Z) . obviously)
+(zero_comp' : ∀ (X : C) {Y Z : C} (f : Y ⟶ Z), (0 : X ⟶ Y) ≫ f = (0 : X ⟶ Z) . obviously)
+
+attribute [instance] has_zero_morphisms.has_zero
+restate_axiom has_zero_morphisms.comp_zero'
+attribute [simp] has_zero_morphisms.comp_zero
+restate_axiom has_zero_morphisms.zero_comp'
+attribute [simp, reassoc] has_zero_morphisms.zero_comp
+
+/-- A category "has a zero object" if it has an object which is both initial and terminal. -/
+class has_zero_object :=
+(zero : C)
+(unique_to : ∀ X : C, unique (zero ⟶ X))
+(unique_from : ∀ X : C, unique (X ⟶ zero))
+
+variables [has_zero_object.{v} C]
+
+variables {C}
+
+instance [has_zero_object.{v} C] : has_zero C :=
+{ zero := has_zero_object.zero.{v} C }
+
+instance unique_from (X : C) : unique (X ⟶ 0) := has_zero_object.unique_from.{v} X
+instance unique_to (X : C) : unique (0 ⟶ X) := has_zero_object.unique_to.{v} X
+
+namespace has_zero_object
+
+def zero_morphism (X Y : C) :=
+inhabited.default (X ⟶ 0) ≫ inhabited.default (0 ⟶ Y)
+
+def hom_has_zero {X Y : C} : has_zero (X ⟶ Y) :=
+{ zero := zero_morphism X Y }
+
+local attribute [instance] hom_has_zero -- in a moment these will be provided by `has_zero_morphism`
+
+lemma zero_morphism_comp {X Y Z : C} (f : Y ⟶ Z) : (0 : X ⟶ Y) ≫ f = (0 : X ⟶ Z) :=
+by { dsimp [has_zero.zero, zero_morphism], rw [category.assoc], congr, }
+lemma comp_zero_morphism {X Y Z : C} (f : X ⟶ Y) : f ≫ (0 : Y ⟶ Z) = (0 : X ⟶ Z) :=
+by { dsimp [has_zero.zero, zero_morphism], rw [←category.assoc], congr, }
+
+-- in a moment these will be deprecated by `has_zero_morphism.zero_comp` and `has_zero_morphism.comp_zero`.
+local attribute [simp] zero_morphism_comp comp_zero_morphism
+
+/-- A category with a zero object has zero morphisms. -/
+instance : has_zero_morphisms.{v} C := {}
```

Comment:
What a lovely proof.

Reply:
There's actually a Lean bug/feature(?) I noticed recently, that prevented me from simplifying quite a few proofs in the limits library to this form. Apparently when you use `auto_param` in a class field, it can turn implicit arguments back into explicit ones... I'll see if I can pin it down.

Reply:
I actually don't really like this style of proof from a software engineering point of view, at least in this case. It's too much "action at a distance". Suppose you want to know how the zero morphisms in your category are defined. Even once you have tracked them down to this `instance : has_zero_morphisms.{v} C`, you still need to figure out why there is no definition of `has_zero` provided and that it comes from an instance 10 lines above.

Much clearer to simply replace everything so far in this section by
```lean
/-- A category with a zero object has zero morphisms. -/
instance : has_zero_morphisms.{v} C :=
{ has_zero := λ X Y, ⟨inhabited.default (X ⟶ 0) ≫ inhabited.default (0 ⟶ Y)⟩,
  comp_zero' := λ X Y f Z, by { dunfold has_zero.zero, rw ←category.assoc, congr, },
  zero_comp' := λ X Y Z f, by { dunfold has_zero.zero, rw category.assoc, congr, }, }
```

Reply:
Indeed, that was horrible. Fixed.